pipeline/lib/metric.ml

```diff
@@ -0,0 +1,313 @@
+     |}
+  in
+  let ( / ) a b = Yojson.Safe.Util.member b a in
+  let prId =
+    Github.Api.exec_graphql api query ~variables >|= fun json ->
+    let open Yojson.Safe.Util in
+    let repo = json / "data" / "repository" in
+    let id = repo / "pullRequest" / "id" in
+    id |> to_string
+  in
+  let _ =
+    prId >|= fun id ->
+    let variables = [ ("subjectId", `String id); ("body", `String text) ] in
+    let mutation =
+      {|
+         mutation ($body: String!, $subjectId: ID!) {
+           addComment(input:{body: $body, subjectId: $subjectId}) {
+             commentEdge {
+               node {
+                 body
+               }
+             }
+           }
+         }
+       |}
+    in
+    Github.Api.exec_graphql api mutation ~variables >|= fun json ->
+    let open Yojson.Safe.Util in
+    let body =
+      json / "data" / "addComment" / "commentEdge" / "node" / "body"
+      |> to_string
+    in
+    print_endline @@ body
+  in
+  ()
+
+module NotifyGithub = struct
+  type t = {
+    conninfo : string;
+    repository : Repository.t;
+    github_app : Github.App.t option;
+    worker : string;
+    docker_image : string;
+    pull_number : int;
+  }
+
+  let id = "notify-github"
+  let pp h (key, _) = Fmt.pf h "notify-github %s" key
+
+  module Key = struct
+    type t = string
+
+    let digest t : string = t
+  end
+
+  module Value = Current.Unit
+  module Outcome = Current.String
+
+  let auto_cancel = true
+
+  let publish
+      { conninfo; repository; worker; docker_image; github_app; pull_number; _ }
+      job _worker_job_id () =
+    let open Lwt.Infix in
+    Current.Job.start job ~level:Current.Level.Above_average >>= fun () ->
+    Logs.info (fun log -> log "Looking for significantly changed metrics...");
+    match
+      Storage.get_main_branch_metrics ~conninfo ~repository ~worker
+        ~docker_image
+    with
+    | Some result_main, Some result -> (
+        let compare_metrics, main_commit =
+          parse_benchmark_data ~data:result_main
+        in
+        let metrics, _ = parse_benchmark_data ~data:result in
+        let changes = find_changed_metrics ~metrics ~compare_metrics in
+        match github_app with
+        | Some app when not (BenchmarksData.is_empty changes) ->
+            let text =
+              format_changes_message ~changes ~repository ~worker ~docker_image
+                ~main_commit
+            in
+            let open Current.Syntax in
+            let _ =
+              let+ installations = Github.App.installations app in
+              let apis = List.map Github.Installation.api installations in
+              let api = match apis with [] -> None | api :: _ -> Some api in
+              match api with
+              | Some api ->
+                  post_github_comment ~api ~repository ~pull_number ~text
+              | None -> ()
+            in
+            Lwt.return (Ok text)
+        | _ -> Lwt.return (Ok "No Github app configured or no changes found"))
```

Comment:
We can probably afford to differentiate between a bad configuration and the lack of changes to ease debugging in the future :P

```ocaml
| Some _ -> Ok "no changes"
| _ -> Error "no github app"
```